homeassistant/components/wiffi/__init__.py

```diff
@@ -0,0 +1,166 @@
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry):
+    """Unload a config entry."""
+    api = hass.data[DOMAIN][config_entry.entry_id]  # type ::= WiffiIntegrationApi
+    await api.server.close_server()
+
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(config_entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        api = hass.data[DOMAIN].pop(config_entry.entry_id)
+        api.shutdown()
+
+    return unload_ok
+
+
+class WiffiIntegrationApi:
+    """API object for wiffi handling. Stored in hass.data."""
+
+    def __init__(self, hass, config_entry):
+        """Initialize the instance."""
+        self._hass = hass
+        self._server = WiffiTcpServer(config_entry.data[CONF_PORT], self)
+        self._known_devices = {}
+        self._async_add_entities = {}
+        self._periodic_callback = async_track_time_interval(
+            hass, self._periodic_tick, timedelta(seconds=10)
+        )
+
+    def shutdown(self):
+        """Shutdown wiffi api.
+
+        Remove listener for periodic callbacks.
+        """
+        remove_listener = self._periodic_callback
+        remove_listener()
+
+    async def __call__(self, device, metrics):
+        """Process callback from by TCP server if new data arives from a device."""
+        if device.mac_address not in self._known_devices:
+            # add all entities of new device
+            self._known_devices[device.mac_address] = {}
+
+            device_info = {
+                "connections": {
+                    (device_registry.CONNECTION_NETWORK_MAC, device.mac_address)
+                },
+                "identifiers": {(DOMAIN, device.mac_address)},
+                "manufacturer": "stall.biz",
+                "name": f"{device.moduletype} {device.mac_address}",
+                "model": device.moduletype,
+                "sw_version": device.sw_version,
+            }
+
+            sensor_entities = []
+            bool_entities = []
+
+            # unique entity id
+            entity_id = device.mac_address.replace(":", "")
+
+            for metric in metrics:
+                entity = None
+                if metric.is_number:
+                    entity = NumberEntity(entity_id, device_info, metric)
+                    sensor_entities.append(entity)
+                elif metric.is_string:
+                    entity = StringEntity(entity_id, device_info, metric)
+                    sensor_entities.append(entity)
+                elif metric.is_bool:
+                    entity = BoolEntity(entity_id, device_info, metric)
+                    bool_entities.append(entity)
+                else:
+                    # unknown type -> ignore
+                    continue
+
+                self._known_devices[device.mac_address][metric.id] = entity
```

Comment:
Don't store the entity. Just keep track of what mac addresses are new and what are existing.

Reply:
done